File: CS50x_Introduction_to_Computer_Science/week_2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int count_letters(const char *text)
{
    int letters = 0;
    for (int i = 0; text[i] != '\0'; i++)
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
    }
    return letters;
}

int count_words(const char *text)
{
    int in_word = 0, words = 0;
    for (int i = 0; text[i] != '\0'; i++)
    {
        if (isspace(text[i]))
        {
            in_word = 0;
        }
        else if (!in_word)
        {
            in_word = 1;
            words++;
        }
    }
    return words;
}

int count_sentences(const char *text)
{
    int sentences = 0;
    for (int i = 0; text[i] != '\0'; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentences++;
        }
    }
    return sentences;
}

int main(void)
{
    char text[1000];

    printf("Text: ");
    fgets(text, sizeof(text), stdin);

    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    float L = (letters * 100.0) / words;
    float S = (sentences * 100.0) / words;

    float index = 0.0588 * L - 0.296 * S - 15.8;
    int grade = round(index);

    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %d\n", grade);
    }

    return 0;
}

File: CS50x_Introduction_to_Computer_Science/week_6/sentimental-readability/readability.py
from cs50 import get_string


def count_letters(text):
    # Count only alphabetic characters
    return sum(1 for c in text if c.isalpha())


def count_words(text):
    # Words are separated by spaces, so words = spaces + 1
    return len(text.split())


def count_sentences(text):
    # Sentences end with ., !, or ?
    return sum(1 for c in text if c in ['.', '!', '?'])


def main():
    text = get_string("Text: ")

    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)

    # L = average number of letters per 100 words
    L = (letters / words) * 100
    # S = average number of sentences per 100 words
    S = (sentences / words) * 100

    # Coleman-Liau index formula
    index = 0.0588 * L - 0.296 * S - 15.8
    grade = round(index)

    if grade < 1:
        print("Before Grade 1")
    elif grade >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {grade}")


if __name__ == "__main__":
    main()
